File: 8. String to Integer (atoi)/atoi.py
__author__ = 'Bill Quan'
"""
Implement atoi to convert a string to an integer.
Hint: Carefully consider all possible input cases. If you want a challenge, please do not see below and ask yourself
      what are the possible input cases.
Notes: It is intended for this problem to be specified vaguely (ie, no given input specs). You are responsible to gather
       all the input requirements up front.
"""


class Solution(object):
    def myAtoi(self, str):
        """
        :param Str: str
        :return: int
        """
        i, sign, ret = 0, 1, 0
        if len(str) == 0:
            return 0
        while i < len(str) and str[i] == ' ':
            i += 1
        if str[i] == '+':
            i += 1
        elif str[i] == '-':
            sign = -1
            i += 1
        while i < len(str) and str[i] >= '0' and str[i] <= '9':
            ret = ret * 10 + ord(str[i]) - ord('0')
            i += 1
        return max(min(ret * sign, 2147483647), -2147483648)

sol = Solution()
print Solution.myAtoi(sol, '1')

File: 8. String to Integer (atoi)/atoi.cpp
using namespace std;
#include <limits>
#include <string>
#include <iostream>

class Solution 
{
public:
	int myAtoi(string str) 
	{
		int sign = 1;
		int rst = 0;
		int i = 0;
		while (i < str.length() && str[i] == ' ') i++;
		if (i >= str.length()) return 0;
		if (str[i] == '-')
		{
			sign = -1;
			i++;
		}
		else if (str[i] == '+')
		{
			i++;
		}
		
		while (i < str.length())
		{
			if (str[i] > '9' || str[i] < '0')
				break;
			// deal with overflow
			if (rst > INT_MAX / 10 || (rst == INT_MAX / 10 && (str[i] - '0') > INT_MAX % 10))
				//|| (sign == 1 && rst * 10 > (INT_MAX - str[i] + '0'))
				//|| (sign == -1 && rst * 10 > (INT_MAX + 1 - str[i] + '0')))
			{
				return sign == 1 ? INT_MAX : INT_MIN;
			}
			rst = rst * 10 + str[i] - '0';
			i++;
		}

		return sign * rst;
	}
};

void main()
{
	Solution sol;
	cout << sol.myAtoi("-2147483648");


	system("pause");
	return;
}
